zio-http/shared/src/main/scala/zio/http/endpoint/CodecErrorHandler.scala

```diff
@@ -0,0 +1,72 @@
+package zio.http.endpoint
+
+import zio.Chunk
+
+import zio.schema.{DeriveSchema, Schema}
+
+import zio.http.Header.Accept.MediaTypeWithQFactor
+import zio.http.codec._
+import zio.http.template.{Dom, body, h1, html, p}
+import zio.http.{MediaType, Response}
+
+final case class CodecErrorHandler[A](
+  log: Option[HttpCodecError => String],
```

Comment:
This has a super weird structure. Usually when I make something with this type of structure, I try to step back and throw away irrelevant details or rethink the abstraction.

First off, I would inline & delete `MapError`, as without it, this class doesn't carry its weight.

Then you have:

```scala
  log: Option[HttpCodecError => String],
  f: (HttpCodecError, Chunk[MediaTypeWithQFactor]) => Err,
  codec: HttpCodec[HttpCodecType.ResponseType, Err],
```

This is still super weird. Picking it apart by capabilities:

1. It MIGHT have the capability to turn an `HttpCodecError` into a `String` (for purposes of logging?).
2. It can turn an `HttpCodecError` into an `Err`.
3. It can turn an `Err` into response, and a response into an `Err`.

A few observations:

 * It seems like we are going from `HttpCodecError` to `Err` only so that we can then go into a response.
 * It is odd to pass `Chunk[MediaTypeWithQFactor]` into the function `f`. Why? Because it must return an `Err`, which is a user-defined error type that, presumably, has no protocol-specific details, such as media type. The only case where this is used seems to be a type-unsafe `Err` (wildcard type used, `_`), which further supports the hypothesis that the media type should not be relevant for `Err`.

Now, let's consider the use case:

> The user wants to encode `HttpCodecError` in a way that is consistent with how they are encoding their own errors, so they can provide a uniform API experience to developers working with the API.

Note that this use case, if correct, is one-sided: there is no requirement for a ZIO HTTP Endpoint client of the API to be able to decode some codec error into a typed error (in fact, it's not really recoverable at that stage).

That means we can probably get away with:

```scala
decodeError: (HttpCodecError, Request) => Response,
domainError: HttpCodec[HttpCodecType.ResponseType, Err],
```

(This allows you to go to string if you want, for logging purposes, inasmuch as Response can be string-ified intelligently.)

On the other hand, if we need the codec errors in the typed error channel, then we might try:

```scala
decodeError: HttpCodec[HttpCodecType.ResponseType, HttpCodecError],
domainError: HttpCodec[HttpCodecType.ResponseType, Err],
```

This one says we can go both ways: from codec errors into responses, and back again; and from typed errors into responses, and back again. The user has control over how to turn `HttpCodecError` into a response, without "contaminating" their own `E` (typed error channel).

This encoding also gets ride of the media types because you can use fallback, perhaps, for different media types.

What do you think?

Reply:
Looking back at this, I see the second formulation does not allow the automatic HTTP client to handle `Err` in a type-safe way. I am not sure if that's a problem because codec errors should not happen.